File: ocaml/AC22.ml
type point = {
    x : int;
    y : int;
    z : int;
  }

type block = {
    id: int;
    inf: point;
    sup: point;
  }

let parse_line id line =
  Scanf.sscanf line "%d,%d,%d~%d,%d,%d" (fun x0 y0 z0 x1 y1 z1 ->
      { id; inf = { x = x0; y = y0; z = z0 }; sup = { x = x1; y = y1; z = z1 }})

let compare_block_z a b =
  Int.compare a.inf.z b.inf.z

let intersect_block_xy a b =
  a.inf.x <= b.sup.x && b.inf.x <= a.sup.x &&
  a.inf.y <= b.sup.y && b.inf.y <= a.sup.y

let rec filter_max accu p blocks =
  match blocks with
  | [] -> accu
  | (hd, _) :: tl ->
     let accu =
       if p hd then
         match accu with
         | [] -> [hd]
         | max :: _ ->
            if max.sup.z < hd.sup.z then
              [hd]
            else if max.sup.z = hd.sup.z then
              hd :: accu
            else
              accu
       else
         accu in
     filter_max accu p tl

let fall_block fallen block =
  let supports = filter_max [] (intersect_block_xy block) fallen in
  let z =
    match supports with
    | [] -> 1
    | hd :: _ -> hd.sup.z + 1 in
  let height = block.sup.z - block.inf.z in
  (
    { id = block.id; inf = { block.inf with z }; sup = { block.sup with z = z + height }},
    supports
  ) :: fallen

let fall_blocks blocks =
  let blocks = List.sort compare_block_z blocks in
  List.fold_left fall_block [] blocks

module IntMap = Map.Make (Int)

module IntSet = Set.Make (Int)

let rec fall_set support_map fallen_blocks =
  let (added, set) =
    IntMap.fold (fun block_id support (added, set) ->
        if not (IntSet.mem block_id set) && not (IntSet.is_empty support) && IntSet.subset support fallen_blocks then
          (true, IntSet.add block_id set)
        else
          (added, set)) support_map (false, fallen_blocks) in
  if added then
    fall_set support_map set
  else
    set

let count_fall support_map id =
  IntSet.cardinal (fall_set support_map (IntSet.singleton id)) - 1

let () =
  let blocks = List.mapi parse_line (List.of_seq (Common.input_lines ())) in
  let blocks = fall_blocks blocks in
  let singletons = List.sort_uniq compare (List.filter_map (fun (_, supports) ->
    match supports with [block] -> Some block.id | _ -> None
  ) blocks) in
  let result_part1 = List.length blocks - List.length singletons in
  Printf.eprintf "Part 1: %d\n" result_part1;
  let support_map = List.fold_left (fun map (block, supports) ->
      IntMap.add block.id (IntSet.of_list (List.map (fun block -> block.id) supports)) map)
    IntMap.empty blocks in
  let result_part2 =
    List.fold_left ( + ) 0 (List.map (count_fall support_map) singletons) in
  Printf.eprintf "Part 2: %d\n" result_part2
  

File: rust/src/bin/ac22.rs
use advent_of_code::{Error, Result, Coords3D};

struct Block {
    inf: Coords3D<u64>,
    sup: Coords3D<u64>,
}

impl Block {
    fn intersect_xy(&self, other: &Block) -> bool {
        self.inf.x <= other.sup.x && other.inf.x <= self.sup.x &&
        self.inf.y <= other.sup.y && other.inf.y <= self.sup.y
    }
}

impl std::str::FromStr for Block {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self> {
        let (inf_str, sup_str) = s.split_once('~').ok_or("'~' missing")?;
        Ok(Self {
            inf: Coords3D::parse(",", inf_str)?,
            sup: Coords3D::parse(",", sup_str)?
        })
    }
}

struct FallenBlock {
    id: usize,
    block: Block,
    support: Vec<usize>,
}

fn filter_support(fallen: &Vec<FallenBlock>, block: &Block) -> (u64, Vec<usize>) {
    let mut max_z = 0;
    let mut max_vec = Vec::new();
    for fallen_block in fallen {
        if !fallen_block.block.intersect_xy(block) {
            continue;
        }
        if max_z < fallen_block.block.sup.z {
            max_z = fallen_block.block.sup.z;
            max_vec.clear();
            max_vec.push(fallen_block.id);
        }
        else if max_z == fallen_block.block.sup.z {
            max_vec.push(fallen_block.id);
        }
    }
    (max_z, max_vec)
}

fn fall_blocks(blocks: &Vec<(usize, Block)>) -> Vec<FallenBlock> {
    let mut fallen = Vec::new();
    for (id, block) in blocks {
        let (z, support) = filter_support(&fallen, &block);
        let z = z + 1;
        let height = block.sup.z - block.inf.z;
        let block = Block {
            inf: Coords3D { z, ..block.inf },
            sup: Coords3D { z: z + height, ..block.sup },
        };
        fallen.push(FallenBlock { id: *id, block, support })
    }
    fallen
}

fn count_fall(
    support: &std::collections::HashMap<usize, std::collections::HashSet<usize>>,
    supported: &std::collections::HashMap<usize, std::collections::HashSet<usize>>,
    id: usize
) -> usize {
    let mut fall_set = std::collections::HashSet::new();
    fall_set.insert(id);
    let mut added_vec: Vec<_> = supported.get(&id).unwrap().iter().collect();
    while let Some(&added) = added_vec.pop() {
        if !fall_set.contains(&added) {
            let support_set = support.get(&added).unwrap();
            if !support_set.is_empty() && support_set.is_subset(&fall_set) {
                fall_set.insert(added);
                if let Some(supported_ids) = supported.get(&added) {
                    added_vec.extend(supported_ids);
                }
            }
        }
    }
    fall_set.len() - 1
}

fn part2(
    blocks: &Vec<FallenBlock>,
    singletons: std::collections::HashSet<usize>
) -> usize {
    let mut support = std::collections::HashMap::new();
    let mut supported = std::collections::HashMap::new();
    for fallen in blocks {
        support.insert(fallen.id, fallen.support.iter().cloned().collect());
        for &support_id in &fallen.support {
            supported.entry(support_id)
                .or_insert(std::collections::HashSet::new())
                .insert(fallen.id);
        }
    }
    singletons.iter().map(|&id| count_fall(&support, &supported, id)).sum()
}

fn main() -> Result<()> {
    let mut blocks: Vec<(usize, Block)> = std::io::stdin().lines()
        .enumerate()
        .map(|(id, line)| -> Result<_> { Ok((id, line?.parse()?)) })
        .collect::<Result<_>>()?;
    blocks.sort_by_key(|(_, block)| block.inf.z);
    let blocks = fall_blocks(&blocks);
    let singletons: std::collections::HashSet<_> = blocks.iter().filter_map(
        |fallen|
        match &fallen.support[..] {
            &[id] => Some(id),
            _ => None,
        }
    ).collect();
    let result_part1 = blocks.len() - singletons.len();
    println!("Part 1: {result_part1}");
    let result_part2 = part2(&blocks, singletons);
    println!("Part 2: {result_part2}");
    Ok(())
}
